### PR TITLE
endpoints: auth between services

### PR DESCRIPTION
**Pull request by DanSanche**
Adding go samples for [endpoints auth between services ](https://cloud.google.com/endpoints/docs/openapi/service-account-authentication)

related to b/109951097 and cl/228405078

### CODE REVIEW

File: endpoints/getting-started/app.go [hunk 1]
```diff
@@ -72,6 +72,7 @@ func (h corsHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	h(w, r)
 }
 
+// [START endpoints_auth_info_backend]
```
Comment (tbpg) [inline endpoints/getting-started/app.go]: Please add a newline after this, before the function comment.

File: endpoints/getting-started/client/main.go [hunk 2]
```diff
@@ -1,23 +1,20 @@
-// Copyright 2016 Google Inc. All rights reserved.
+// Copyright 2018 Google LLC. All rights reserved.
```
Comment (tbpg) [inline endpoints/getting-started/client/main.go]: No need to update. Also, see #757.

File: endpoints/getting-started/client/main.go [hunk 3]
```diff
@@ -26,150 +23,103 @@ import (
...
-		resp, err = doEcho()
-	} else if *serviceAccount != "" {
-		resp, err = doJWT()
-	}
+	jwt, err := generateJWT(*serviceAccountFile, *serviceAccountEmail, *audience, 3600)
 	if err != nil {
 		log.Fatal(err)
 	}
-	b, err := httputil.DumpResponse(resp, true)
+	fmt.Println(jwt)
```
Comment (tbpg) [inline endpoints/getting-started/client/main.go]: Should this have any extra context? Something like `fmt.Printf("JWT: %+v", jwt)`? Maybe `"%v"` if you don't want field names.

File: endpoints/getting-started/client/main.go [hunk 4]
```diff
@@ -26,150 +23,103 @@ import (
...
-func doEcho() (*http.Response, error) {
-	msg := map[string]string{
-		"message": *echo,
+// [START endpoints_generate_jwt_sa]
+
+// generateJWT creates a signed JSON Web Token using a Google API Service Account.
+func generateJWT(saKeyfile, saEmail, audience string, expiryLength int64) (string, error) {
+	now := time.Now().Unix()
+
+	// build payload
```
Comment (tbpg) [inline endpoints/getting-started/client/main.go]: Either delete or capitalize + add a period.

File: endpoints/getting-started/client/main.go [hunk 5]
```diff
@@ -26,150 +23,103 @@ import (
...
+		Algorithm: "RS256",
+		Typ:       "JWT",
 	}
-	return http.Post(*host+"/echo?key="+*apiKey, "application/json", &buf)
-}
 
-// doJWT performs an authenticated request using the credentials in the service account file.
-func doJWT() (*http.Response, error) {
-	sa, err := ioutil.ReadFile(*serviceAccount)
+	// sign with keyfile
```
Comment (tbpg) [inline endpoints/getting-started/client/main.go]: Either delete or capitalize + add a period. If commenting, might want more than this below. It's hard to understand what's happening in each step.

File: endpoints/getting-started/client/main.go [hunk 6]
```diff
@@ -26,150 +23,103 @@ import (
...
-	jwsHeader := &jws.Header{
-		Algorithm: "RS256",
-		Typ:       "JWT",
-	}
+// [START endpoints_jwt_request]
 
-	msg, err := jws.Encode(jwsHeader, jwt, rsaKey)
+// makeJWTRequest sends an authorized request to your deployed endpoint.
+func makeJWTRequest(signedJWT, url string) (string, error) {
+	client := &http.Client{}
```
Comment (tbpg) [inline endpoints/getting-started/client/main.go]: Might want to add a timeout.

File: endpoints/getting-started/client/main.go (hunk 3 above)
Comment (DanSanche) [reply endpoints/getting-started/client/main.go]: I decided to just remove it. It's encrypted data, so not particularly useful beyond debugging anyway

**Comment by DanSanche** [comment]
@tbpg I addressed your comments, let me know if there's anything else

File: endpoints/getting-started/client/main.go [hunk 7]
```diff
@@ -1,4 +1,4 @@
-// Copyright 2018 Google LLC. All rights reserved.
+// Copyright 2016 Google Inc. All rights reserved.
```
Comment (tbpg) [inline endpoints/getting-started/client/main.go]: Need to update this license following #757.

File: endpoints/getting-started/client/main.go [hunk 8]
```diff
@@ -100,8 +100,10 @@ func generateJWT(saKeyfile, saEmail, audience string, expiryLength int64) (strin
 // [START endpoints_jwt_request]
 
 // makeJWTRequest sends an authorized request to your deployed endpoint.
-func makeJWTRequest(signedJWT, url string) (string, error) {
-	client := &http.Client{}
+func makeJWTRequest(signedJWT, url string, timeout time.Duration) (string, error) {
+	client := &http.Client{
+		Timeout: timeout,
```
Comment (tbpg) [inline endpoints/getting-started/client/main.go]: Not sure this needs to be an argument to the function - `10 * time.Second` is probably good. Up to you.

Comment (DanSanche) [reply endpoints/getting-started/client/main.go]: yeah, that sounds easier

**Pull request merged**